### PR TITLE
Fix merge / conflicts / resolve for V2.

### PR DESCRIPTION
**Pull request by olsen232**
## Description

Moved decode_path into dataset superclass, which was enough to get `sno merge`, `sno conflicts` and `sno resolve` working.

Conversely, moved `cast_primary_key` out of datasets superclass and into v1 - datasets V2 has its own, schema.sanitise_pk_values, and right now these are not part of the public interface.

## Related links:

https://github.com/koordinates/sno/issues/72

## Checklist:

- [x] Have you reviewed your own change?
- [x] Have you included test(s)?
- [x] Have you updated the [changelog](https://github.com/koordinates/sno/blob/master/CHANGELOG.md)?

### CODE REVIEW

File: sno/structure.py [hunk 1]
```diff
@@ -87,15 +87,15 @@ def version(self):
...
+        Given a path in the sno repository - eg "path/to/dataset/.sno-table/49/3e/Bg==" -
         returns a tuple in either of the following forms:
-        1. (table, "feature", primary_key)
-        2. (table, "meta", metadata_file_path)
+        1. (dataset_path, "feature", primary_key)
+        2. (dataset_path, "meta", meta_item_path)
         """
-        table, table_path = path.split("/.sno-table/", 1)
-        return (table,) + self.get(table).decode_path(table_path)
+        dataset_path, rel_path = re.split("(?=/.sno-table/)", full_path, 1)
```
Comment (craigds) [inline sno/structure.py:97]: should pre-compile this regex outside the function body. Also it should be a raw string

File: sno/structure.py [hunk 2]
```diff
@@ -275,6 +275,9 @@ def instantiate(cls, tree, path, version):
         if not isinstance(tree, pygit2.Tree):
             raise TypeError(f"Expected Tree object, got {type(tree)}")
 
+        if not ".sno-table" in tree:
```
Comment (craigds) [inline sno/structure.py:278]:
```suggestion
        if ".sno-table" not in tree:
```
my linter would dislike this immensely :)

File: sno/structure.py (hunk 1 above)
Comment (olsen232) [reply sno/structure.py:97]: Timed this - precompiling a regex is 25% faster, using simple string split and then re-prepending it is 75% faster. Using simple string operations now. No guarantee that it makes any difference to the code overall, but no harm in it.

File: sno/structure.py (hunk 2 above)
Comment (olsen232) [reply sno/structure.py]: Done

**Pull request merged**